File: T-Bank/Algorithms-and-Data-Structures-2025/11sprint-Graph2/contest/A-Система_непересекающихся_множеств/solution.cpp
#include <iostream>
#include <string>
#include <vector>


class UnionFind {
private:
    int n;
    std::vector<int> parent;
    std::vector<int> rank;
    std::vector<int> minValue;
    std::vector<int> maxValue;

public:
    UnionFind(int n_) : n(n_) {
        parent.resize(n, 0);
        rank.resize(n, 1);
        minValue.resize(n, 0);
        maxValue.resize(n, 0);
        for (int i = 0; i < n; ++i) {
            parent[i] = i;
            minValue[i] = i;
            maxValue[i] = i;
        }
    }
    
    int find(int x) {
        while (x != parent[x]) {
            parent[x] = parent[parent[x]];
            x = parent[x];
        }
        return x;
    }
    
    void union_(int v, int u) {
        int vParent = find(v), uParent = find(u);
        if (vParent == uParent) return;
        if (rank[vParent] >= rank[uParent]) {
            parent[uParent] = vParent;
            rank[vParent] += rank[uParent];
            minValue[vParent] = std::min(minValue[vParent], minValue[uParent]);
            maxValue[vParent] = std::max(maxValue[vParent], maxValue[uParent]);
        } else {
            parent[vParent] = uParent;
            rank[uParent] += rank[vParent];
            minValue[uParent] = std::min(minValue[uParent], minValue[vParent]);
            maxValue[uParent] = std::max(maxValue[uParent], maxValue[vParent]);
        }
    }
    
    void get(int x) {
        int xParent = find(x);
        std::printf("%d %d %d\n", minValue[xParent], maxValue[xParent], rank[xParent]);
    }
};



void solution() {
    /*
    Time Complexity O(MlogN)
    Memory Complexity O(N)
    */
    int n, m;
    std::scanf("%d %d", &n, &m);
    UnionFind uf(n + 1);
    for (int i = 0; i < m; ++i) {
        std::string op;
        std::cin >> op;
        if (op == "union") {
            int u, v;
            std::scanf("%d %d", &u, &v);
            uf.union_(u, v);
        } else {
            int x;
            std::scanf("%d", &x);
            uf.get(x);
        }
    }
}


int main() {
    solution();
}

File: T-Bank/Algorithms-and-Data-Structures-2025/11sprint-Graph2/contest/A-Система_непересекающихся_множеств/solution.py
import sys


class UnionFind:
    def __init__(self, n: int) -> None:
        self.parent: list[int] = list(range(n))
        self.rank: list[int] = [1] * n
        self.min_value: list[int] = list(range(n))
        self.max_value: list[int] = list(range(n))
    
    def find(self, x: int) -> int:
        while x != self.parent[x]:
            self.parent[x] = self.parent[self.parent[x]]
            x = self.parent[x]
        return x
    
    def union_(self, u: int, v: int) -> None:
        u_parent: int = self.find(u)
        v_parent: int = self.find(v)
        if u_parent == v_parent: return
        if self.rank[u_parent] >= self.rank[v_parent]:
            self.parent[v_parent] = u_parent
            self.rank[u_parent] += self.rank[v_parent]
            self.min_value[u_parent] = min(self.min_value[u_parent], self.min_value[v_parent])
            self.max_value[u_parent] = max(self.max_value[u_parent], self.max_value[v_parent])
        else:
            self.parent[u_parent] = v_parent
            self.rank[v_parent] += self.rank[u_parent]
            self.min_value[v_parent] = min(self.min_value[v_parent], self.min_value[u_parent])
            self.max_value[v_parent] = max(self.max_value[v_parent], self.max_value[u_parent])
    
    def get(self, x: int) -> None:
        x_parent: int = self.find(x)
        sys.stdout.write('{} {} {}\n'.format(self.min_value[x_parent], self.max_value[x_parent], self.rank[x_parent]))


def solution() -> None:
    """
    Time Complexity O(MlogN)
    Memory Complexity O(N)
    """
    n, m = map(int, sys.stdin.readline().rstrip().split())
    uf: UnionFind = UnionFind(n + 1)
    for _ in range(m):
        args: list[str] = sys.stdin.readline().rstrip().split()
        if args[0] == 'union':
            uf.union_(*map(int, args[1:]))
        else:
            uf.get(int(args[1]))


if __name__ == '__main__':
    solution()